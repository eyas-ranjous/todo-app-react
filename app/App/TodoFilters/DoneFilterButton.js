import React from 'react';
import { useSelector } from 'react-redux';
import classNames from 'classnames';
import styles from './styles.css';

const DoneFilterButton = ({ setFilter }) => {
  const TodoStatus = useSelector((state) => state.TodoStatus);
  const TodoFilters = useSelector((state) => state.TodoFilters);
  const selectedFilter = useSelector((state) => state.selectedFilter);

  const filter = TodoFilters.DONE;
  const statusList = [TodoStatus.DONE];

  const classes = classNames(
    styles['filter-button'],
    styles['filter-button-done'],
    selectedFilter === filter ? styles['filter-button-selected'] : ''
  );

  return (
    <button
      className={classes}
      onClick={() => setFilter(statusList, filter)}
    >
      Done
    </button>
  );
};

export default DoneFilterButton;
