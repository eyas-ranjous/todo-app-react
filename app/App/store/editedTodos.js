const editedTodos = (state = new Set(), action) => {
  if (action.type === 'EDIT_TODO') {
    return new Set(state).add(action.todoId);
  }

  if (action.type === 'REMOVE_EDITED_TODO') {
    const newState = new Set(state);
    newState.delete(action.todoId);
    return newState;
  }

  return state;
};

export default editedTodos;
