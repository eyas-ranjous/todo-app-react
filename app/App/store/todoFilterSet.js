const todoFilterSet = (state = new Set(), action) => {
  if (action.type === 'UPDATE_STATUS_FILTER_SET') {
    return new Set(action.status);
  }

  return state;
};

export default todoFilterSet;
