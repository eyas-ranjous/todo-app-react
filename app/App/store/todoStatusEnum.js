export const TodoStatusEnum = {
  TODO: 1,
  DONE: 2
};

export const TodoStatus = (state = TodoStatusEnum) => state;
