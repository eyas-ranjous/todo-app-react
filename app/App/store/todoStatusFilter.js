const todoStatusFilter = (state = new Set(), action) => {
  if (action.type === 'SET_TODO_STATUS_FILTER') {
    return new Set(action.status);
  }

  return state;
};

export default todoStatusFilter;
