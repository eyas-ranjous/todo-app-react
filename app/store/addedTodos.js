import { TodoStatusEnum } from './todoStatusEnum';

let nextTodoId = 1;

const addedTodos = (state = new Map(), action) => {
  if (action.type === 'ADD_TODO') {
    const { todo } = action;
    const id = `todo-${nextTodoId}`;
    const status = TodoStatusEnum.TODO;
    nextTodoId += 1;
    return new Map(state).set(id, { ...todo, ...{ id, status } });
  }

  if (action.type === 'UPDATE_TODO') {
    const { todo } = action;

    if (!state.has(todo.id)) return state;

    const updatedTodo = { ...state.get(todo.id), ...todo };
    return new Map(state).set(todo.id, updatedTodo);
  }

  if (action.type === 'REMOVE_TODO') {
    const newState = new Map(state);
    newState.delete(action.todoId);
    return newState;
  }

  if (action.type === 'TOGGLE_TODO_STATUS') {
    if (!state.has(action.todoId)) return state;

    const todo = state.get(action.todoId);
    const { DONE, TODO } = TodoStatusEnum;
    const newStatus = todo.status === TODO ? DONE : TODO;
    const updatedTodo = { ...todo, ...{ status: newStatus } };

    return new Map(state).set(todo.id, updatedTodo);
  }

  return state;
};

export default addedTodos;
