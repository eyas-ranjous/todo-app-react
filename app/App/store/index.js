import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatus';
import todoStatusFilter from './todoStatusFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  todoStatusFilter
});

export default createStore(reducers);
