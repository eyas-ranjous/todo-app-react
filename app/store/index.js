import { createStore, combineReducers } from 'redux';
import addedTodos from './addedTodos';
import editedTodos from './editedTodos';
import { TodoStatus } from './todoStatusEnum';
import { TodoFilters } from './todoFiltersEnum';
import todoFilterSet from './todoFilterSet';
import selectedFilter from './selectedFilter';

const reducers = combineReducers({
  addedTodos,
  editedTodos,
  TodoStatus,
  TodoFilters,
  todoFilterSet,
  selectedFilter
});

export default createStore(reducers);
