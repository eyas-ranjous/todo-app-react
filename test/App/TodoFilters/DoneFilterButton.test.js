import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import { TodoFiltersEnum } from '../../../app/store/TodoFiltersEnum';
import { TodoStatusEnum } from '../../../app/store/TodoStatusEnum';
import createStore from '../../../app/store';
import DoneFilterButton from '../../../app/App/TodoFilters/DoneFilterButton';

describe('DoneFilterButton', () => {
  const store = createStore();

  it('renders filter button', () => {
    const setFilter = jest.fn();
    const wrapper = mount(
      <Provider store={store}>
        <DoneFilterButton setFilter={setFilter} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.find('button').simulate('click');
    expect(setFilter)
      .toHaveBeenCalledWith([TodoStatusEnum.DONE], TodoFiltersEnum.DONE);
  });

  it('renders filter button being selected', () => {
    const setFilter = jest.fn();
    store.dispatch({
      type: 'SET_SELECTED_STATUS_FILTER',
      filter: TodoFiltersEnum.DONE
    });
    const wrapper = mount(
      <Provider store={store}>
        <DoneFilterButton setFilter={setFilter} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
