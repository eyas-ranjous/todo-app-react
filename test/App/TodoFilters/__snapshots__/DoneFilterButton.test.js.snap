// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoneFilterButton renders filter button 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <DoneFilterButton
    setFilter={[MockFunction]}
  >
    <button
      className="filter-button filter-button-done"
      onClick={[Function]}
    >
      Done
    </button>
  </DoneFilterButton>
</Provider>
`;

exports[`DoneFilterButton renders filter button being selected 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <DoneFilterButton
    setFilter={[MockFunction]}
  >
    <button
      className="filter-button filter-button-done filter-button-selected"
      onClick={[Function]}
    >
      Done
    </button>
  </DoneFilterButton>
</Provider>
`;
